### DISCUSSION TITLE
Add to cart Button Collection Pages

### DISCUSSION THREAD

**Discussion by mircomartin** [EvgeniyMukhamedjanov/liquid-ajax-cart#57] [General]
@EvgeniyMukhamedjanov I would like to add the button my products grid! How can i do that? I've just createad this scripts, works fine, but the carts wont update until the refresh of browser.

```
const productsCard = document.querySelectorAll('.product-card.grids');

if ( productsCard.length > 0) {
		
	for (let i = 0; i < productsCard.length; i++) {

		const element = productsCard[i];
		const buttonAddToCart = element.querySelector('#addtocartGrid')
		
			buttonAddToCart.addEventListener('submit', e => {
				e.preventDefault()
				
				const idProductElement = element.querySelector('[name="id"]');
				const idProduct = idProductElement.getAttribute('value')
	
				let formData = {
					'items': [{
					 'id': idProduct,
					 'quantity': 1
					 }]
				 };
				 fetch(window.Shopify.routes.root + 'cart/add.js', {
					 method: 'POST',
					 headers: {
						 'Content-Type': 'application/json'
					 },
					 body: JSON.stringify(formData)
				 })
				 .then(response => {
					 return response.json();
				 })
				 .catch((error) => {
					 console.error('Error:', error);
				 });

			})

		}

}
```

**Comment by EvgeniyMukhamedjanov** [comment]
As I mentioned [here](https://github.com/EvgeniyMukhamedjanov/liquid-ajax-cart/discussions/55), if you want to add a product to the cart using JavaScript, you should use the Liquid Ajax Cart API rather than `fetch` in order to simultaneously update all the cart related elements on the page.

**Reply by mircomartin** [reply]
@EvgeniyMukhamedjanov yes! i've just created this script, but it does not work!

```
function clickHandler(e) {
  e.preventDefault()
	var plus = element.querySelector('[name="id"]').value;

  liquidAjaxCart.cartRequestAdd({ 
    items: [
      {
        id: Number(plus),
        quantity: 1
      }
    ]  
  }, {} )
}

document.addEventListener('DOMContentLoaded', function () {
		const productsCard = document.querySelectorAll('.product-card.grids');
		const addToCartGrid = document.querySelectorAll('.product-card.grids #addtocartGrid button');
		// do something here ...
		
		liquidAjaxCart.subscribeToCartSectionsUpdate(sections => {
			sections.forEach(section => {
				section.elements.forEach(element => {
					element.querySelectorAll('.product-card.grids #addtocartGrid button').forEach(plus => {
						plus.addEventListener('click', clickHandler);
					})
				})
			})
		})


	}, false);
```

**Reply by mircomartin** [reply]
@EvgeniyMukhamedjanov i've just modified my script now works! but cart does not updated can you check it? 
```
function clickHandler(element) {
	var plus = element.querySelector('[name="id"]').value;
  liquidAjaxCart.cartRequestAdd({ 
    items: [
      {
        id: Number(plus),
        quantity: 1
      }
    ]  
  }, {} )
}

document.addEventListener('DOMContentLoaded', function () {

	const productsCard = document.querySelectorAll('.product-card.grids');
	// do something here ...
	
	for (let i = 0; i < productsCard.length; i++) {
		const element = productsCard[i];
		element.addEventListener('click', () => {
			console.log('click')
			clickHandler(element)

		});
		
	}
	
}, false);
```

**Reply by EvgeniyMukhamedjanov** [reply]
Share a preview link pls, i'll check

**Reply by mircomartin** [reply]
@EvgeniyMukhamedjanov yes! you can see here! https://mm-c-theme.myshopify.com/collections/all - password "mirco" , the add to cart works but the problem is the cart-drawer it wont update until refresh browser

**Reply by EvgeniyMukhamedjanov** [reply]
You need to have a parent element with the `data-ajax-cart-section` attribute for the empty cart as well. Liquid Ajax Cart doesn't know what element to update without it.

**Reply by mircomartin** [reply]
@EvgeniyMukhamedjanov mmm i really dont understand, i shoud add that tag data tag in my grid?

**Reply by EvgeniyMukhamedjanov** [reply]
You use the `data-ajax-cart-section` attribute only for the cart items. So there is no this attribute when the cart is empty.
<img width="1280" alt="image" src="https://user-images.githubusercontent.com/118430645/206401456-1cf28002-e77e-4074-8c8c-054532316996.png">

It is better to use this attribute for the common parent element of the empty cart and not empty cart. Maybe the element with the `bg-white drawer__wrapper` class is a good fit for it. Try to move the `data-ajax-cart-attribute` to that element.

**Reply by mircomartin** [reply]
@EvgeniyMukhamedjanov ohh nice! Works now! the script was fine?

**Reply by EvgeniyMukhamedjanov** [reply]
Should be fine

**Reply by mircomartin** [reply]
@EvgeniyMukhamedjanov what is the best place for put the cart tags? in the head? or before close body? what is your recomendation?

**Reply by EvgeniyMukhamedjanov** [reply]
Somewhere after the footer is good

**Reply by mircomartin** [reply]
@EvgeniyMukhamedjanov for example if the add to cart fail because i don have stock, how can i show a message? I mean how can i get the error?
For example here in the function of shopify have error, but here liquidAjaxCart.cartRequestAdd() im really lost!
 ```
fetch('/cart/add.js', {
      method: 'POST',
      credentials: 'same-origin',
      headers: {
        'Content-Type': 'application/json',
        'X-Requested-With': 'XMLHttpRequest'
      },
      body: JSON.stringify(addData)
    }).then(function (data) {
      if (data.status == 200) {
        window.location.reload;
        return
      }
      else {
        console.log('Request returned an error', data)
      };
    })
    .catch(function (error) {
      console.log('Request failed', error);
    });
```

**Reply by EvgeniyMukhamedjanov** [reply]
You can use the `lastComplete` callback as described here: https://liquid-ajax-cart.js.org/reference/cartRequestAdd/
```javascript
liquidAjaxCart.cartRequestAdd({ 
  items: [
    {
      id: Number(plus),
      quantity: 1
    }
  ]  
}, {
  lastComplete: requestState => {
    console.log(requestState.responseData)
  }
})
```

**Reply by mircomartin** [reply]
@EvgeniyMukhamedjanov I have a gift wrapping in the cart, the problem is i need not count that product, i've just hide in the cart with liquid loop, but the quantiy is there. can you test it here https://shl2gj87tf7vb6tu-29614735444.shopifypreview.com, right now i dont use the ajax cart count for the numner, i use without ajax.
